retrieval/scrape.go

```diff
@@ -0,0 +1,243 @@
+package retrieval
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/log"
+	"golang.org/x/net/context"
+
+	"github.com/prometheus/prometheus/storage"
+)
+
+// Scraper retrieves samples.
+type Scraper interface {
+	// Sends samples to ch until an error occurrs, the context is canceled
+	// or no further samples can be retrieved. The channel is closed before returning.
+	Scrape(ctx context.Context, ch chan<- *model.Sample) error
+}
+
+// Scrape pools runs scrapers for a set of targets where a target, identified by
+// a label set is scraped exactly once.
+type ScrapePool struct {
+	appender storage.SampleAppender
+
+	ctx    context.Context
+	cancel func()
+
+	mtx    sync.RWMutex
+	loops  map[model.Fingerprint]*scrapeLoop
+	states map[model.Fingerprint]*TargetStatus
+}
+
+func NewScrapePool(app storage.SampleAppender) *ScrapePool {
+	sp := &ScrapePool{
+		appender: app,
+		states:   map[model.Fingerprint]*TargetStatus{},
+	}
+	sp.ctx, sp.cancel = context.WithCancel(context.Background())
+
+	return sp
+}
+
+func (sp *ScrapePool) Stop() {
+	if sp.cancel != nil {
+		sp.cancel()
+	}
+}
+
+// Sync the running scrapers with the provided list of targets.
+func (sp *ScrapePool) Sync(targets []*Target) {
+	sp.mtx.Lock()
+	defer sp.mtx.Unlock()
+
+	// Shutdown all running loops after their current scrape
```

Comment:
The calls to `Sync()` only happen in one second intervals by the target manager if there were any updates. So yes, that's "worse" in theory than before. Bugs are worse.